### ISSUE TITLE
Since Ansible 2.3.0, Powershell modules do not honor ANSIBLE_KEEP_REMOTE_FILES

### ISSUE THREAD

**Issue by trondhindenes** [ansible/ansible#23986]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
core

##### ANSIBLE VERSION
```
2.3.0
```

##### CONFIGURATION
export ANSIBLE_KEEP_REMOTE_FILES=1

##### SUMMARY
Using Ansible 2.3.0, generated ps1 files are not longer left on the target node when setting the env var ANSIBLE_KEEP_REMOTE_FILES to "1" on Windows-based target nodes.
As far as I can see the folders in %temp% are left, but the actual files inside are deleted, except for files executed using 'script'. This makes it next to impossible to troubleshoot/debug during module development.

**Comment by trondhindenes** [comment]
Pinging @nitzmahone on this one since I'm currently stuck without this functionality :-|. I've tried looking at $myInvocation inside the script by hacking powershell.ps1 but there's essentially nothing there.

**Comment by nitzmahone** [comment]
Yeah, the persistence path didn't get cleaned up in time for 2.3.0 (since I was flogging the become-on-ntlm/kerb issue), so basically it's "forced pipeline" mode all the time with the new exec wrapper.

On the plus side, there is no more module replacer stuff happening, so you can just run powershell.ps1 once to load the module API, then run the module code directly (set the preparsed arg dict in a global $complex_args). The debug story actually gets much better on this codepath. So as long as you just want to debug your module (ie, you're not having trouble with the exec wrapper itself), you can just use ISE with the raw source files. Debugging the wrapper is currently a major PITA that requires a special harness, though I'm hoping to simplify a few things going forward to make it easier (minimizing the use of runspaces, etc).

**Comment by trondhindenes** [comment]
ok thanks. Just a bit of a new workflow then - I can work with that :-)
I'm keeping this issue open as I assume this still is a bug (that the files arent left in place, that is)

**Comment by trondhindenes** [comment]
Thinking a bit more about this, the most important part of debugging is often to see the actual vars passed in to Ansible. But I can hack that using the $complex_args variable and just stream those out to disk while waiting for this to get addressed.

**Comment by trondhindenes** [comment]
Sorry to bug this one down, but we need to come up with some guidance, detailed walkthru or something here. I'm struggling, which means that the majority of devs working on ps-based modules are struggling. I had to:
1. Hack powershell.ps1 to emit complex_args into a json file with this code:
```
$FileName = "ansible_complex_args_$($complex_args._ansible_module_name)_$([System.Guid]::NewGuid().Guid).json"
Set-Content -Path (Join-Path $env:temp "ansiblecomplex_args.json") -Value ($complex_args | convertto-json) -force
```
2. copy powershell.ps1 onto my target node and running to import the statement (after disabling the export-modulemember code since that is only valid when run from inside a module)
3. copy the problematic module onto my target node

Still I can't get it working - I tried simply parsing complex args using
`$Complex_Args = get-content "C:\Users\<<account>>\AppData\Local\Temp\ansiblecomplex_args.json" | ConvertFrom-Json`
And then run my module code, which still fails - probably from how I populate `complex_args`, but I can't see how to fix it.

I'd appreciate some detailed guidance here. IMHO we should be very tuned to making sure module devs can be as productive as possible, and while I have no trouble understanding that major changes are sometimes required, it seems to me that we forgot to provide an alternative method of doing efficient debugging. It could be as simple as a blog post or an update to Ansible docs, but we should have _something_. I'm still stuck.

**Comment by jhawkesworth** [comment]
Just to say I chatted with Matt about this a bit a the windows working group yesterday so he's aware.- -meeting transcript is here - probably better than me paraphrasing it: https://meetbot.fedoraproject.org/ansible-meeting/2017-04-28/windows_working_group.2017-04-28-16.30.log.html

**Comment by cwegener** [comment]
I just got hit by this hard. I was pulling my hair out.
I'm using the win_dsc5 module from @trondhindenes and I need to have a look at how this new exec pipeline impacts the win_dsc5 module.
In terms of ANSIBLE_KEEP_REMOTE_FILES, I'm dead in the water since I have no .ps1 files to run. The .ps1 files get created dynamically.

**Comment by nitzmahone** [comment]
Here's a quick description of the way I usually do module debug:
- Get module_utils/powershell.ps1 and the module code on a Windows host. I usually create symlinks in my Linux host's vagrant dir to module_utils and modules/windows in my Ansible checkout, but do whatever you like to get those things accessible.
- Run ISE, open powershell.ps1 and the module file, then F5 powershell.ps1 (only need to do this once per session to get the Ansible shared module code defined- in 2.3+ you'll see an error about Export-ModuleMember- ignore it).
- In the ISE console, set `$complex_args` to the module args you want to use (in Powershell dict format) , eg, for `win_ping`:
```
$complex_args = @{data="blarg"}
```
- Set breakpoints as necessary, then F5 the module code in ISE, do whatever other debug/stepping/hacking you need. In my case, since I'm hacking the original module code, then I just go do my staging/commits back in Git, push, lather-rinse-repeat. :)

So long as you don't need to debug the wrapper itself, this process works great (and should also work for previous Ansible versions). If you *do* need to debug the wrapper, well, that's what'll require the stuff I'm still working on.

**Comment by trondhindenes** [comment]
Thanks for chiming in @nitzmahone .
@cwegener what I had to do was to hack powershell.ps1 as described above to get to the complex_args contents, and then convert that json into a dict that I can then pass to the module script itself (see the function ConvertPSObjectToHashtable here: http://stackoverflow.com/questions/22002748/hashtables-from-convertfrom-json-have-different-type-from-powershells-built-in-h)

For me it is extremely important to be able to get to the "real" value of the complex_args, as much troubleshooting revolves around unexpected parameters being sent in (we're using dynamic vars throughout our playbooks, so it may hit us extra hard). So imho, any debug workflow that relies on the operator/dev "knowing" the contents of complex_args is not sufficient for our troubleshooting needs.

Again, I know this is on the list already, but I want to make sure that we're all aligned on the "problem statement", so to speak.

**Comment by cwegener** [comment]
Thanks @nitzmahone for the detailed steps. I'll try them with win_dsc5 and I'll see what I get.

@trondhindenes I think I understand what you're saying. I'll have to do some more in-depth soulsearching with win_dsc5 :smile: So far I've only scratched the surface of win_dsc5 but it is my favorite module.

On a side note on win_dsc5, I found out the other day that psrunascredential_username and psrunascredential_password simply work like magic in win_dsc5 because of your existing credentials logic. Very well done on the design of that module!

**Comment by cwegener** [comment]
I successfully used @nitzmahone procedure for my bug triage. Turns out the win_dsc5 module and ansible is not to blame in my current problem case that I'm working on.

In my case was able to find a bug in a community DSC resource that was causing my problems. The problem even occurred when using the DSC resource without any win_dsc5.

The instructions from @nitzmahone were a major part in getting my triage back on the right track though.

So, to summarize my story, my triage process was thrown off completely by the fact that ANSIBLE_KEEP_REMOTE_FILES doesn't get honored in 2.3 for Windows targets at the moment.  I probably could have moved a lot faster through my triage if ANSIBLE_KEEP_REMOTE_FILES would still work for Windows targets in 2.3.
